consensus/polybft/validator_set.go

```diff
@@ -91,27 +104,80 @@ type validatorSet struct {
 
 	// totalVotingPower denotes voting power of entire validator set
 	totalVotingPower int64
+
+	// quorum
+	quorumSize uint64
+
+	// votingPowerMap represents voting powers per validator address
+	votingPowerMap map[types.Address]uint64
+
+	// logger instance
+	logger hclog.Logger
 }
 
 // NewValidatorSet creates a new validator set.
-func NewValidatorSet(valz AccountSet) (*validatorSet, error) {
+func NewValidatorSet(valz AccountSet, logger hclog.Logger) (*validatorSet, error) {
 	var validators = make([]*ValidatorAccount, len(valz))
 	for i, v := range valz {
 		validators[i] = NewValidator(v, 0)
 	}
 
 	validatorSet := &validatorSet{
 		validators: validators,
+		logger:     logger.Named("validator_set"),
 	}
 
-	err := validatorSet.updateWithChangeSet()
+	// populate voting power map
+	validatorSet.populateVotingPower()
+
+	// calculate quorum according to submitted voting powers
+	err := validatorSet.calculateQuorum()
+	if err != nil {
+		return nil, err
+	}
+
+	err = validatorSet.updateWithChangeSet()
 	if err != nil {
 		return nil, fmt.Errorf("cannot update changeset: %w", err)
 	}
 
 	return validatorSet, nil
 }
 
+// populateVotingPower populates voting power map
+// for each validator in the current validator set
+func (v *validatorSet) populateVotingPower() {
+	v.votingPowerMap = make(map[types.Address]uint64, len(v.validators))
+	for _, validator := range v.validators {
+		v.votingPowerMap[validator.Metadata.Address] = validator.Metadata.VotingPower
+	}
+}
+
+// calculateQuorum calculates quorum size for given voting power map
+func (v *validatorSet) calculateQuorum() error {
+	totalVotingPower := uint64(0)
+	for _, v := range v.votingPowerMap {
+		totalVotingPower += v
+	}
+
+	if totalVotingPower == 0 {
+		return errInvalidTotalVotingPower
+	}
+
+	// If there cannot be faulty nodes (less than 4 nodes in the network),
+	// then quorum size is determined as total voting power (namely all the nodes must send vote).
+	// Otherwise quorum size is calculated as 2/3 supermajority
+	if v.calcMaxFaultyNodes() == 0 {
```

Comment:
This confuses me. Why do we need it?

Reply:
I followed existing implementation on the Edge (it basically requires that for small networks, less than 4 nodes where we don't have BFT properties met, all the validators need to send a message in order to have a quorum):
https://github.com/0xPolygon/polygon-edge/blob/feature/v3-parity/consensus/ibft/validators.go#L38

However we can remove it if you don't find it necessary.

Reply:
If there are 3 or less validators then according to 
```
func (v validatorSet) calcMaxFaultyNodes() uint64 {
	return uint64((v.Len() - 1) / 3)
}
```
number of faulty validators is 0. If that is the case, all of the validators must vote positively, so the qourum size is equal to total voting power.

Reply:
I've removed it, so the code is simpler and behavior consistent: [daae347](https://github.com/0xPolygon/polygon-edge/pull/904/commits/daae347f6e768e6e825c349d673706504c9610e1).